File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/bukkit/pipeline/ItemCloseOnClickInterceptor.java
package me.devnatan.inventoryframework.bukkit.pipeline;

import me.devnatan.inventoryframework.api.VirtualView;
import me.devnatan.inventoryframework.api.component.Component;
import me.devnatan.inventoryframework.api.component.ItemComponent;
import me.devnatan.inventoryframework.api.pipeline.PipelineContext;
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor;
import me.devnatan.inventoryframework.bukkit.context.SlotClickContext;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.jetbrains.annotations.NotNull;

/**
 * Intercepted when a player clicks on an item the view container. Checks if the container should be
 * closed when the item is clicked.
 */
public final class ItemCloseOnClickInterceptor implements PipelineInterceptor<VirtualView> {

	@Override
	public void intercept(@NotNull PipelineContext<VirtualView> pipeline, @NotNull VirtualView subject) {
		if (!(subject instanceof SlotClickContext)) return;

		final SlotClickContext context = (SlotClickContext) subject;
		final InventoryClickEvent event = context.getClickOrigin();
		if (event.getSlotType() == InventoryType.SlotType.OUTSIDE) return;

		final Component component = context.getComponent();
		if (!(component instanceof ItemComponent) || !component.isVisible()) return;

		final ItemComponent item = (ItemComponent) component;
		if (item.isCloseOnClick()) {
			context.closeForPlayer();
			pipeline.finish();
		}
	}
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/minestom/pipeline/ItemCloseOnClickInterceptor.kt
package me.devnatan.inventoryframework.minestom.pipeline

import me.devnatan.inventoryframework.api.VirtualView
import me.devnatan.inventoryframework.api.component.ItemComponent
import me.devnatan.inventoryframework.api.pipeline.PipelineContext
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor
import me.devnatan.inventoryframework.minestom.context.SlotClickContext
import net.minestom.server.event.inventory.InventoryPreClickEvent

/**
 * Intercepted when a player clicks on an item the view container. Checks if the container should be
 * closed when the item is clicked.
 */
class ItemCloseOnClickInterceptor : PipelineInterceptor<VirtualView> {
    override fun intercept(
        pipeline: PipelineContext<VirtualView>,
        subject: VirtualView,
    ) {
        if (subject !is SlotClickContext) return

        val event: InventoryPreClickEvent = subject.clickOrigin
        event.inventory ?: return

        val component = subject.component
        if (component !is ItemComponent || !component.isVisible) return

        val item: ItemComponent = component
        if (item.isCloseOnClick) {
            subject.closeForPlayer()
            pipeline.finish()
        }
    }
}
